refactor: Preallocate result in TryParseHex to avoid resizing

l0rinc opened the pull request:
This pull request introduces optimizations to the `TryParseHex` function, focusing primarily on the ideal case (valid hexadecimal input without spaces).
A new benchmark, `HexParse` was introduced in a separate commit.

The main optimization preallocates the result vector based on the input string's length. This aims to completely avoid costly dynamic reallocations when no spaces are present.

------------

Before:
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.60 |      623,238,893.11 |    0.3% |      0.01 | `HexParse`
|                1.65 |      606,747,566.34 |    0.6% |      0.01 | `HexParse`
|                1.60 |      626,149,544.07 |    0.3% |      0.01 | `HexParse`
```

After:
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.68 |    1,465,555,976.27 |    0.8% |      0.01 | `HexParse`
|                0.68 |    1,472,962,920.18 |    0.3% |      0.01 | `HexParse`
|                0.68 |    1,476,159,423.00 |    0.3% |      0.01 | `HexParse`
```

Commit:
Add benchmark for TryParseHex

Running `make && ./src/bench/bench_bitcoin -filter=HexParse` a few times results in:
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.60 |      623,238,893.11 |    0.3% |      0.01 | `HexParse`
|                1.65 |      606,747,566.34 |    0.6% |      0.01 | `HexParse`
|                1.60 |      626,149,544.07 |    0.3% |      0.01 | `HexParse`
```

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -81,16 +81,21 @@ template <typename Byte>
     std::vector<Byte> vch;
-    auto it = str.begin();
-    while (it != str.end()) {
-        if (IsSpace(*it)) {
```
we don't expect most hex strings to contain spaces so let's fall back to this when we can't parse the string instead

paplorinc replied:
Valid question, that's why I put this change in a separate commit with before/after measurements, resulting in an additional 30% speedup on my machine:
[Before:](https://github.com/bitcoin/bitcoin/pull/29458/commits/e467437dea8b78e87ba16af306304c7740d5815c)
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.69 |    1,445,181,075.98 |    0.5% |      0.01 | `TryParseHexBenchmark`
|                0.68 |    1,459,941,407.03 |    0.2% |      0.01 | `TryParseHexBenchmark`
|                0.69 |    1,457,349,650.60 |    1.1% |      0.01 | `TryParseHexBenchmark`
```
[After:](https://github.com/bitcoin/bitcoin/pull/29458/commits/a37a5b11bfba8e8cd2c5f9e940770b6b844ecedc)
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.54 |    1,866,216,380.08 |    0.9% |      0.01 | `TryParseHexBenchmark`
|                0.53 |    1,871,755,872.12 |    0.5% |      0.01 | `TryParseHexBenchmark`
|                0.53 |    1,875,106,332.14 |    0.1% |      0.01 | `TryParseHexBenchmark`
```

andrewtoth replied:
Since I'm here, tried this as well. On my machine the speedup is not so severe, but still measurable:

Preallocation only (e467437dea8b78e87ba16af306304c7740d5815c)
|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                0.84 |    1,186,108,418.00 |    0.3% |           15.21 |            1.86 |  8.172 |           3.37 |    0.0% |      0.53 | `TryParseHexBenchmark`

Skipping whitespace check as well (a37a5b11bfba8e8cd2c5f9e940770b6b844ecedc)
|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                0.76 |    1,320,747,604.66 |    0.1% |           14.24 |            1.67 |  8.521 |           2.87 |    0.0% |      0.54 | `TryParseHexBenchmark`

andrewtoth replied:
Hmm thinking about this more, maybe saving 0.08ns is not worth this churn in the codebase. I think I agree with @maflcko that this isn't necessary and preallocation on its own is a simpler change with bigger benefit.

paplorinc replied:
Thanks guys, the code is indeed cleaner without this, I've rebased the PR, please check it again

paplorinc replied:
Since we've left the `IsSpace` calls here (which were checking the ASCII whitespaces one-by-one for each converted character), I've optimized it in a separate PR to short circuit for non-whitespace cases, see: https://github.com/bitcoin/bitcoin/pull/29602

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -81,16 +81,21 @@ template <typename Byte>
-        auto c1 = HexDigit(*(it++));
-        if (it == str.end()) return std::nullopt;
-        auto c2 = HexDigit(*(it++));
-        if (c1 < 0 || c2 < 0) return std::nullopt;
```
c1 validation can be done sooner

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -81,16 +81,21 @@ template <typename Byte>
 std::optional<std::vector<Byte>> TryParseHex(std::string_view str)
 {
     std::vector<Byte> vch;
```
Ideally (when no spaces present) two characters will form a single byte - and when spaces are present, it will be less, so we can avoid reallocation by reserving `str.size() / 2`

Empact replied:
nit: may be unnecessary / obvious, but an inline comment saying // two hex characters form a single byte will be more helpful to readers of the code than the comment you've added above.

paplorinc replied:
Done, thanks

andrewtoth commented on src/util/strencodings.cpp:
```diff
@@ -81,16 +81,21 @@ template <typename Byte>
+    vch.reserve(str.size() / 2);
+
+    for (auto it = str.begin(); it != str.end(); ++it) {
+        auto c1 = HexDigit(*it);
```
nit: `const auto c1{HexDigit(*it)};`

paplorinc replied:
Thanks, done

andrewtoth commented on src/util/strencodings.cpp:
```diff
@@ -81,16 +81,21 @@ template <typename Byte>
+            }
+        } else if (++it == str.end()) return std::nullopt;
+
+        auto c2 = HexDigit(*it);
```
nit: `const auto c2{HexDigit(*it)};`

andrewtoth commented on src/bench/parse_hex.cpp:
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2009-2024 The Bitcoin Core developers
```
```suggestion
// Copyright (c) 2024- The Bitcoin Core developers
```

andrewtoth reviewed (approved):
ACK a37a5b11bfba8e8cd2c5f9e940770b6b844ecedc

Bench before:
|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.88 |      532,130,184.93 |    1.4% |           29.43 |            4.13 |  7.118 |           6.40 |    0.0% |      0.53 | `TryParseHexBenchmark`

After:
|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                0.82 |    1,226,337,041.45 |    2.5% |           14.24 |            1.80 |  7.931 |           2.87 |    0.0% |      0.54 | `TryParseHexBenchmark`

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -81,16 +81,21 @@ template <typename Byte>
+            } else {
+                return std::nullopt;
+            }
+        } else if (++it == str.end()) return std::nullopt;
```
The style is a bit all over the place (I tried to adhere to the previous style), should I add braces everywhere in this method?

andrewtoth replied:
It took me a minute to figure out where `it` was being incremented when you put it in the `else if`. Not a big deal but maybe add a break here and move this check above where we define `c2` so it's more clear that is where we are moving `it` forward?

```suggestion
        }
        
        if (++it == str.end()) return std::nullopt;
```

andrewtoth replied:
I don't think you need braces for a return early though.

paplorinc replied:
Rebased the last commit to keep the history clean. 
hanks for taking the time to review this PR thoroughly, appreciate it.

andrewtoth commented on src/util/strencodings.cpp:
```diff
@@ -1,5 +1,5 @@
 // Copyright (c) 2009-2010 Satoshi Nakamoto
-// Copyright (c) 2009-2022 The Bitcoin Core developers
+// Copyright (c) 2024- The Bitcoin Core developers
```
I think you changed this copyright notice in the wrong file. I don't think this one should be changed, only the one in the new file `parse_hex.cpp`.

paplorinc replied:
Got it, fixed, thanks

paplorinc commented on src/test/span_tests.cpp:
```diff
@@ -53,7 +53,7 @@ BOOST_AUTO_TEST_SUITE(span_tests)
 //
 // Previously there was a bug where writing a SFINAE check for vector<bool> was
-// not possible, because in libstdc++ vector<bool> has a data() memeber
+// not possible, because in libstdc++ vector<bool> has a data() member
```
completely unrelated, but got this warning from the CI

andrewtoth replied:
Hmm not sure we should have these changes in an unrelated PR. The CI failure was for a whitespace issue in the hex parsing code.

paplorinc replied:
Moved it to a separate PR https://github.com/bitcoin/bitcoin/pull/29471

paplorinc commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -52,7 +52,7 @@ class WalletStorage
     virtual bool HasEncryptionKeys() const = 0;
     virtual bool IsLocked() const = 0;
-    //! Callback function for after TopUp completes containining any scripts that were added by a SPKMan
+    //! Callback function for after TopUp completes containing any scripts that were added by a SPKMan
```
same

andrewtoth commented:
Your last commit contains changes that should instead should be applied to the first and third commit and then rebased.

andrewtoth commented on src/util/strencodings.cpp:
```diff
@@ -81,16 +81,19 @@ template <typename Byte>
     std::vector<Byte> vch;
-    auto it = str.begin();
-    while (it != str.end()) {
+    vch.reserve(str.size() / 2);
```
I think for the cleanest diff the second commit could only have just this one line, no?

paplorinc replied:
Done, thanks

Commit:
Preallocate result in `TryParseHex` to avoid resizing

Running `make && ./src/bench/bench_bitcoin -filter=HexParse` a few times results in:
```
|           ns/base16 |            base16/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.68 |    1,465,555,976.27 |    0.8% |      0.01 | `HexParse`
|                0.68 |    1,472,962,920.18 |    0.3% |      0.01 | `HexParse`
|                0.68 |    1,476,159,423.00 |    0.3% |      0.01 | `HexParse`
```

Empact commented:
Concept ACK - I think you could do without the 3rd commit.

paplorinc commented:
Thanks for the review @Empact>
We could do without, that's why it's in a separate commit. It's very low risk however and makes the code more readable. If you think maintainability is risky, I can add more tests - what do you think?

paplorinc commented:
I have received contradictory comments, so I've removed the last commit to simplify this change by avoiding the resize, hoping this satisfies everyone.

Empact commented:
> We could do without, that's why it's in a separate commit. It's very low risk however and makes the code more readable. If you think maintainability is risky, I can add more tests - what do you think?

@paplorinc note that Bitcoin Core is under different constraints than a typical software project. We are review-constrained and change-skeptical in part because we have to be on guard against adversarial changes and also because we are appreciative of the importance of the codebase to the Bitcoin ecosystem.

That said, refactoring and cleanup are important to the long-term health of the project. One way to be respectful of review time is to make focused, well-motivated pull requests, that do not mix various motivations in the same PR.

See also: https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#refactoring

Empact reviewed (approved):
Code review ACK https://github.com/bitcoin/bitcoin/pull/29458/commits/0d127ca5ebedb5a1eee7b183be06545d05236090

andrewtoth reviewed (approved):
ACK 0d127ca5ebedb5a1eee7b183be06545d05236090

Before:

|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.76 |      567,923,390.61 |    0.6% |           29.43 |            3.88 |  7.577 |           6.40 |    0.0% |     10.95 | `HexParse`

After:
|           ns/base16 |            base16/s |    err% |      ins/base16 |      cyc/base16 |    IPC |     bra/base16 |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                0.81 |    1,234,838,738.87 |    0.5% |           15.21 |            1.78 |  8.534 |           3.37 |    0.0% |     10.84 | `HexParse`

hebasto commented on src/bench/parse_hex.cpp:
```diff
@@ -0,0 +1,33 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <bench/bench.h>
+#include <util/strencodings.h>
+#include <random.h>
+#include <cassert>
```
nit: https://api.cirrus-ci.com/v1/task/5912720797597696/logs/ci.log:
```
bench/parse_hex.cpp should add these lines:
#include <stddef.h>             // for size_t
#include <optional>             // for operator==, nullopt, optional
#include <vector>               // for vector

bench/parse_hex.cpp should remove these lines:

The full include-list for bench/parse_hex.cpp:
#include <bench/bench.h>        // for Bench, doNotOptimizeAway, BENCHMARK, PriorityLevel
#include <random.h>             // for FastRandomContext
#include <stddef.h>             // for size_t
#include <util/strencodings.h>  // for allocator, string, TryParseHex, char_traits
#include <cassert>              // for assert
#include <optional>             // for operator==, nullopt, optional
#include <vector>               // for vector
---

hebasto reviewed (commented):
Concept ACK 0d127ca5ebedb5a1eee7b183be06545d05236090.

Observing ~x2 speed up on my machine.

hebasto reviewed (approved):
ACK a19235c14b3dc02de30b5d769de29d1752c23dbd.

andrewtoth reviewed (approved):
Re-ACK a19235c14b3dc02de30b5d769de29d1752c23dbd

Empact commented:
Re-ACK https://github.com/bitcoin/bitcoin/pull/29458/commits/a19235c14b3dc02de30b5d769de29d1752c23dbd

Empact reviewed (approved):
Re-ACK https://github.com/bitcoin/bitcoin/pull/29458/commits/a19235c14b3dc02de30b5d769de29d1752c23dbd

achow101 commented:
ACK a19235c14b3dc02de30b5d769de29d1752c23dbd
